XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+                (m_outputHeight + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchY
+                1 
+            };
+
+            // EASU/RCAS common
+            utilities::shader::Defines defines;
+            defines.add("FSR_THREAD_GROUP_SIZE", 64);
+            defines.add("SAMPLE_SLOW_FALLBACK", 1);
+            defines.add("SAMPLE_BILINEAR", 0);
+            defines.add("SAMPLE_HDR_OUTPUT", 0);
+
+            // EASU specific
+            defines.add("SAMPLE_RCAS", 0);
+            defines.add("SAMPLE_EASU", 1);
+            m_shaderEASU = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR EASU CS", threadGroups, defines.get(), shadersDir.string());
+
+            // RCAS specific
+            defines.set("SAMPLE_EASU", 0);
+            defines.set("SAMPLE_RCAS", 1);
+            m_shaderRCAS = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR RCAS CS", threadGroups, defines.get(), shadersDir.string());
+
+            m_isSharpenOnly = false;
+        }
+
+        void initializeSharpen() {
+            initializeScaler();
+
+            // TODO
+
+            //m_isSharpenOnly = true;
+        }
+
+        void initializeIntermediary(uint32_t width, uint32_t height, int64_t format) {
+            if (format == 0) {
+                // good balance between visuals and perf
+                format = m_device->getTextureFormat(TextureFormat::R16G16B16A16_UNORM);
+            }
+
+            DebugLog("FSRUpscaler initializeIntermediary with %u, %u, %u\n", width, height, format);
+
+            if (m_device->isTextureFormatSRGB(format)) {
+                format = DXGI_FORMAT_R8G8B8A8_UNORM;
+                DebugLog("  sRGB output format changed to: %u\n", format);
+            }
+
+            // create the intermediary texture between upscale and sharpen pass
+            XrSwapchainCreateInfo info;
+            ZeroMemory(&info, sizeof(info));
+            info.width = width;
+            info.height = height;
+            info.format = format;
+            info.arraySize = 1;
+            info.mipCount = 1;
+            info.sampleCount = 1;
+            info.usageFlags = XR_SWAPCHAIN_USAGE_SAMPLED_BIT | XR_SWAPCHAIN_USAGE_UNORDERED_ACCESS_BIT;
+            m_intermediary = m_device->createTexture(info, "FSR Intermediary TEX2D", 0, 0, nullptr);
+        }
+
+        const std::shared_ptr<IConfigManager> m_configManager;
+        const std::shared_ptr<IDevice> m_device;
+        const uint32_t m_outputWidth;
+        const uint32_t m_outputHeight;
+
+        uint32_t m_inputWidth;
+        uint32_t m_inputHeight;
+        bool m_isSharpenOnly{false};
+
+        std::shared_ptr<IComputeShader> m_shaderEASU;
+        std::shared_ptr<IComputeShader> m_shaderRCAS;
+        std::shared_ptr<IShaderBuffer> m_configBuffer;
+        std::shared_ptr<ITexture> m_intermediary;
```

Comment:
Not an issue for now, but for DX12 I wonder if this texture can be unique for the processing chain, or it needs to be added to the triple-buffered texture chain.

Reply:
We might need to revisit the processing chain resources differently later. This one is only a dependency due to FSR 2 passes processing though.